task/net/src/vlan.rs

```diff
@@ -0,0 +1,407 @@
+    /// Requests that a packet waiting in the rx queue of `socket` be delivered
+    /// into loaned memory at `payload`.
+    ///
+    /// If a packet is available and fits, copies it into `payload` and returns
+    /// its `UdpMetadata`. Otherwise, leaves `payload` untouched and returns an
+    /// error.
+    fn recv_packet(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        socket: SocketName,
+        payload: idol_runtime::Leased<idol_runtime::W, [u8]>,
+    ) -> Result<UdpMetadata, RequestError<NetError>> {
+        let socket_index = socket as usize;
+
+        if generated::SOCKET_OWNERS[socket_index].0.index()
+            != msg.sender.index()
+        {
+            return Err(NetError::NotYours.into());
+        }
+
+        // Iterate over all of the per-VLAN sockets, returning the first
+        // available packet with a bonus `vid` tag attached in the metadata.
+        for (i, vid) in VLAN_RANGE.enumerate() {
+            let socket = self.get_socket_mut(socket_index, i)?;
+            match socket.recv() {
+                Ok((body, endp)) => {
+                    if payload.len() < body.len() {
+                        return Err(RequestError::Fail(ClientError::BadLease));
+                    }
+                    payload
+                        .write_range(0..body.len(), body)
+                        .map_err(|_| RequestError::went_away())?;
+
+                    return Ok(UdpMetadata {
+                        port: endp.port,
+                        size: body.len() as u32,
+                        addr: endp.addr.try_into().map_err(|_| ()).unwrap(),
+                        vid,
+                    });
+                }
+                Err(smoltcp::Error::Exhausted) => {
+                    // Keep iterating
+                }
+                Err(_) => {
+                    // uhhhh TODO
+                    // (keep iterating in the meantime)
+                }
+            }
+        }
+        Err(NetError::QueueEmpty.into())
+    }
+
+    /// Requests to copy a packet into the tx queue of socket `socket`,
+    /// described by `metadata` and containing the bytes loaned in `payload`.
+    fn send_packet(
+        &mut self,
+        msg: &userlib::RecvMessage,
+        socket: SocketName,
+        metadata: UdpMetadata,
+        payload: idol_runtime::Leased<idol_runtime::R, [u8]>,
+    ) -> Result<(), RequestError<NetError>> {
+        let socket_index = socket as usize;
+        if generated::SOCKET_OWNERS[socket_index].0.index()
+            != msg.sender.index()
+        {
+            return Err(NetError::NotYours.into());
+        }
+
+        // Convert from absolute VID to an index in our VLAN array
+        if !VLAN_RANGE.contains(&metadata.vid) {
+            return Err(NetError::InvalidVLan.into());
+        }
+        let vlan_index = metadata
+            .vid
+            .checked_sub(VLAN_RANGE.start)
+            .ok_or(NetError::InvalidVLan)?;
```

Comment:
Could this be a normal subtraction instead, given the `VLAN_RANGE.contains(..)` check that precedes it?

Reply:
Yes, good catch – this `checked_sub` predated `VLAN_RANGE`